lxd/device/disk.go

```diff
@@ -2540,55 +2540,32 @@ func (d *disk) generateVMConfigDrive() (string, error) {
 
 	instanceConfig := d.inst.ExpandedConfig()
 
-	// Use an empty vendor-data file if no custom vendor-data supplied.
-	vendorDataKey := "cloud-init.vendor-data"
-	vendorData, ok := instanceConfig[vendorDataKey]
-	if !ok {
-		vendorDataKey = "user.vendor-data"
-		vendorData = instanceConfig[vendorDataKey]
-		if vendorData == "" {
-			vendorData = "#cloud-config\n{}"
-		}
+	// Get raw data from instance config.
+	vendorDataKey := cloudinit.GetEffectiveConfigKey(instanceConfig, "vendor-data")
+	userDataKey := cloudinit.GetEffectiveConfigKey(instanceConfig, "user-data")
+	vendorData, userData := cloudinit.GetResultingCloudConfig(instanceConfig, vendorDataKey, userDataKey, d.inst.Name(), d.inst.Project().Name)
```

Comment:
In my view it would be clearer to return a `cloudinit.Config` (or similarly named) struct that contained fields for userdata, vendordata etc rather than multiple return values as it allows you to pass around the overall cloud config.